### ISSUE TITLE
SK translation/ no module message

### ISSUE THREAD

**Issue by lylax47** [czcorpus/kontext#1086]
Apologies for the long absence. So I have two new issues that I can't seem to fix, seemingly unrelated. The first is the kontext log:
----- 
2016-04-26 20:46:36,099 [translation] WARNING: Failed to load translations for sk_SK with error: IOError(2, 'No translation file found for domain')
2016-04-26 20:46:36,108 [controller] ERROR: CorpInfoNotFound (/var/lib/mantee/registry/susanne)
Traceback (most recent call last):
  File "/var/www/kontext/public/../lib/controller.py", line 785, in run
    self._pre_action_validate()
  File "/var/www/kontext/public/../lib/controller.py", line 498, in _pre_action_validate
    raise err
CorpInfoNotFound: CorpInfoNotFound (/var/lib/mantee/registry/susanne)

2016-04-26 20:46:36,108 [QUERY] INFO: {"user_id": 0, "proc_time": 0.0016, "settings": {"ctxattrs": "word", "shuffle": 1, "pagesize": 40, "wlpagesize": 25, "kwicrightctx": "10", "attrs": "word", "cup_hl": "q", "structs": "p,g,err,corr", "citemsperpage": 50, "kwicleftctx": "-10"}, "pid": 4485, "request": {"HTTP_USER_AGENT": "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:45.0) Gecko/20100101 Firefox/45.0", "REMOTE_ADDR": "127.0.0.1"}, "action": "message", "params": {}, "user": "anonymous", "date": "2016-04-26 20:46:36.108603"}
The second is in the apache logs, and states that there is no module message. As I understand, it is searching for the message.tmpl file, which I have located under kontext/templates/ . 
------
Tue Apr 26 20:46:36 2016] [error] [client 127.0.0.1] mod_wsgi (pid=4485): Exception occurred processing WSGI script '/var/www/kontext/public/app.py'.
[Tue Apr 26 20:46:36 2016] [error] [client 127.0.0.1] Traceback (most recent call last):
[Tue Apr 26 20:46:36 2016] [error] [client 127.0.0.1]   File "/var/www/kontext/public/app.py", line 203, in __call__
[Tue Apr 26 20:46:36 2016] [error] [client 127.0.0.1]     status, headers, sid_is_valid, body = app.run(request)
[Tue Apr 26 20:46:36 2016] [error] [client 127.0.0.1]   File "/var/www/kontext/public/../lib/controller.py", line 825, in run
[Tue Apr 26 20:46:36 2016] [error] [client 127.0.0.1]     self.output_result(methodname, tmpl, result, action_metadata, outf=output)
[Tue Apr 26 20:46:36 2016] [error] [client 127.0.0.1]   File "/var/www/kontext/public/../lib/controller.py", line 965, in output_result
[Tue Apr 26 20:46:36 2016] [error] [client 127.0.0.1]     template_class = self._get_template_class(template[:-5])
[Tue Apr 26 20:46:36 2016] [error] [client 127.0.0.1]   File "/var/www/kontext/public/../lib/controller.py", line 383, in _get_template_class
[Tue Apr 26 20:46:36 2016] [error] [client 127.0.0.1]     tpl_file, pathname, description = imp.find_module(name, ['%s/..' % template_dir])
[Tue Apr 26 20:46:36 2016] [error] [client 127.0.0.1] ImportError: No module named message

**Comment by tomachalek** [comment]
There are probably several issues involved here:

## 1.
Remove the following *language* element from your *config.xml* (*kontext/global/translations/language[text()="sk_SK"]*)

## 2.

Install the *susanne* corpus. Although it is not required it can make the installation easier as *susanne* is (for historical reasons) used as kind of a fallback corpus. Sources of the corpus can be downloaded [here](http://corpora.fi.muni.cz/noske/src/example-corpora/susanne-example-source.tar.bz2). There is also a *Makefile* in the package which should make the compilation (indexing) quite easy.

After moving the index files into a destination directory a so called *registry file* must be created. It contains information about corpus structure and data location. It should be named after the corpus (here: *susanne*) and reside in a directory specified in *config.xml* (*kontext/corpora/manatee_registry*). The contents are:

```
PATH "/path/to/a/directory/containing/susanne/indices"
ENCODING utf-8

ATTRIBUTE word
ATTRIBUTE lemma
ATTRIBUTE tag
ATTRIBUTE status

STRUCTURE doc {
        ATTRIBUTE file
        ATTRIBUTE n
}

STRUCTURE p
STRUCTURE head {
        ATTRIBUTE type
}

STRUCTURE font {
        ATTRIBUTE type
}

STRUCTURE g
```

## 3.

The 'missing *message* template' error could have been caused by one of the two errors above but I will try to investigate it further. It may be just a problem during an error processing.

**Comment by lylax47** [comment]
Well that's the problem. We have already installed the susanne corpus, and the registry file is located in the path where it says it can not find it. Is it necessary for the registry file and the vert file to be located within the kontext folder?

**Comment by tomachalek** [comment]
Looking at the error message *ERROR: CorpInfoNotFound (/var/lib/mantee/registry/susanne)*
I can see a typo in the registry path: "mantee" instead of *manatee*

**Comment by lylax47** [comment]
Well I feel stupid. Thanks.

**Comment by tomachalek** [comment]
You're welcome. Typo errors are usually easier to spot for others than for an author :)

**Comment by lylax47** [comment]
So I'm getting a similar error as before. The "No module named message" seems to have solved itself, but now it can't find the template first_form, which is in the same folder as message.tmpl. It's the same error, just "No module name first_form" at the end.

**Comment by lylax47** [comment]
Sorry I realize now it's not looking for these files. It seems to not be correctly compiling the tmpl files. Under kontext/public/files/js/test I have a firstform.js, but if I'm not mistaken this should be in the .../tpl folder?

**Comment by tomachalek** [comment]
Please look into the *cmpltmpl* directory and in case there is only a single file *__init__.py* it means that the HTML templates are not compiled (the *templates* directory contains just sources for these compiled ones).

Please note that KonText must be (kind of) built before it can be used. We use [Grunt.js](http://gruntjs.com/) to configure and perform all the necessary tasks. If you have a working [Node.js](https://nodejs.org/en/) on you computer/server then you just run 

```
npm install; grunt production
```
or 
```
npm install; grunt devel
```
(the devel version makes less JS/CSS optimizations)

In general, this action compiles all the HTML templates, merges/transpiles miscellaneous JS/TS/CSS/... files, configures required plug-ins and moves the production/testing-ready files into proper directories.

More information can be found here: https://github.com/czcorpus/kontext/blob/master/INSTALL.md

In case you have already done this then I will need more information (a related piece of application.log).

**Comment by lylax47** [comment]
Ok, so first off no, there is only one __init__.py file. I've tried several things, but to no avail. I've reinstalled grunt and nodejs. When I try to run grunt in kontext/ it says that it can not find task production, even though the task is clearly created at the end of gruntfile.js.

**Comment by lylax47** [comment]
jselm:  ucnkQueryStorage
Loading "Gruntfile.js" tasks...ERROR
>> TypeError: Cannot read property 'textContent' of undefined
Warning: Task "production" not found. Use --force to continue.

**Comment by tomachalek** [comment]
It looks like you are using a little bit outdated version from the *master* branch (according to the log record "jselm: ucnkQueryStorage") - can you please try to pull the latest changes first? Because I have made some changes in *Gruntfile.js* (and in a related file *scripts/grunt/kontext.js*) which may help. Even if it won't help, it will be easier to figure it out with the current version.

**Comment by lylax47** [comment]
##### So I pulled the newest branch(rel 0.8), and tried running grunt production after npm install and got a similar error again:
Loading "Gruntfile.js" tasks...ERROR
>> TypeError: kontext.getCustomStyles is not a function
Warning: Task "production" not found. Use --force to continue.

Aborted due to warnings.

**Comment by tomachalek** [comment]
It looks like some kind of a version mix-up. *Gruntfile.js* calls *getCustomStyles()* only in the recent development version (*master* branch). But this new version also contains updated *scripts/grunt/kontext.js* which contains a required implementation.

Please check that you have no unwanted local changes (git status) and if it is not a problem for you, please try the "master" branch which will became the 0.9 version in a few weeks anyway. (But the version 0.8 should work too)

**Comment by lylax47** [comment]
##### Ok recloned the master version, everything worked fine up until requirejs:production. There seems to be a file it is searching for which does not exist.

Running "requirejs:production" (requirejs) task
>> Error: Error: ENOENT: no such file or directory, open '/var/www/kontext/public/files/js/min/plugins/applicationBar/init.js'
>> In module tree:
>>     tpl/coll
>>       tpl/document
>> 
>>     at Error (native)
Warning: Task "requirejs:production" failed. Use --force to continue.

Aborted due to warnings.

**Comment by tomachalek** [comment]
Just append *<application_bar />* and *<footer_bar />* elements to '<plugins>'. These are optional plug-ins which can be turned off but they still must be present in *config.xml*. I should really update that *config.sample.xml* - sorry for that.

**Comment by lylax47** [comment]
Sorry I don't fully understand, which elements I need to append to *config.xml*.

**Comment by tomachalek** [comment]
```xml
<kontext>
...
<plugins>
  <application_bar />
  <footer_bar />
  <!-- ...rest of the plug-ins definitions ... -->
</plugins>
...
</kontext>
```

**Comment by lylax47** [comment]
Thanks, just figured it out as you messaged me! )))

**Comment by lylax47** [comment]
##### I'm very sorry for all of these questions. I'm hoping that this is one of the last ones. So grunt compiled everything fine, but now I'm getting another error in apache logs. Application logs just prints JSON details.

[Thu Apr 28 18:17:27 2016] [error] [client 127.0.0.1] Traceback (most recent call last):
[Thu Apr 28 18:17:27 2016] [error] [client 127.0.0.1]   File "/var/www/kontext/public/app.py", line 206, in __call__
[Thu Apr 28 18:17:27 2016] [error] [client 127.0.0.1]     status, headers, sid_is_valid, body = app.run()
[Thu Apr 28 18:17:27 2016] [error] [client 127.0.0.1]   File "/var/www/kontext/public/../lib/controller.py", line 823, in run
[Thu Apr 28 18:17:27 2016] [error] [client 127.0.0.1]     self.output_result(methodname, tmpl, result, action_metadata, outf=output)
[Thu Apr 28 18:17:27 2016] [error] [client 127.0.0.1]   File "/var/www/kontext/public/../lib/controller.py", line 966, in output_result
[Thu Apr 28 18:17:27 2016] [error] [client 127.0.0.1]     result.respond(CheetahResponseFile(outf))
[Thu Apr 28 18:17:27 2016] [error] [client 127.0.0.1]   File "/var/www/kontext/public/../cmpltmpl/document.py", line 1661, in respond
[Thu Apr 28 18:17:27 2016] [error] [client 127.0.0.1]     _v = VFFSL(SL,"join_params",False)(VFFSL(SL,"q",True), VFFSL(SL,"globals",True), [('fcrit', VFFSL(SL,"fcrit_shortref",True)), ('ml', 0)]) # u"$join_params($q, $globals, [('fcrit', $fcrit_shortref), ('ml', 0)])" on line 300, col 111
[Thu Apr 28 18:17:27 2016] [error] [client 127.0.0.1]   File "/var/www/kontext/public/../lib/templating/__init__.py", line 96, in join_params
[Thu Apr 28 18:17:27 2016] [error] [client 127.0.0.1]     for k, v in a])
[Thu Apr 28 18:17:27 2016] [error] [client 127.0.0.1]   File "/var/www/kontext/public/../lib/templating/__init__.py", line 85, in quote
[Thu Apr 28 18:17:27 2016] [error] [client 127.0.0.1]     return werkzeug.urls.url_quote(s, unsafe='+')
[Thu Apr 28 18:17:27 2016] [error] [client 127.0.0.1] TypeError: url_quote() got an unexpected keyword argument 'unsafe'

**Comment by tomachalek** [comment]
No problem :) I'm aware that the current installation process is far from perfect or easy.

The error message you've passed here looks like a version problem related to the *werkzeug* package (in older versions, the method *url_quote* does not support the *unsafe* arg). Please run a Python interactive shell (just type *python* in a console) and write:

```python
import werkzeug
print(werkzeug.__version__)
```

If you see anything smaller than 0.10.x (e.g. 0.9.x) you have to upgrade the package. The process depends on how you've installed the current one. In case it is installed via your Linux distribution package manager (e.g. _apt-get install python-werkzeug_ in Ubuntu) then the easiest way is to remove the distribution package (_apt-get remove python-werkzeug_ in Ubuntu) and use an alternative installed via Python package manager *pip* (in case you don't have one you must install it via you distribution - _apt-get install python-pip_ in Ubuntu):

```
pip install Werkzeug
```

If you already have a version of the *werkzeug* package installed via *pip* then type:

```
pip install --upgrade Werkzeug
```

**Comment by lylax47** [comment]
Once again, impeccable timing, haha. I just got i working. I had upgraded werkzeug before I messaged you, and apparently all I needed to restart apache... Thank you so much for all your help! If you'd like I'll be writing up installation instructions in Russian, but I could write them in English as well for future reference, and/or I could send you the minimalist config.xml file I ended up with.

**Comment by tomachalek** [comment]
:)

Having installation instructions in Russian would be great - I'm all for it. As regards the English version -  I have to review the current INSTALL.md file and my notes written in Czech first - but I am ready to review/accept pull requests for English docs too. Hopefully I will be able to fetch some information from your future notes as I can read Russian a bit (yet my writing and speaking is much worse).

I have created a new directory *doc* for the installation instructions (and possibly other documentation). Please name the Russian instructions file *INSTALL.ru.md*. Once it is ready, just create a pull request and I will merge it.

Thank you for the offer and feel free to send the pull request at any time you want (no need to rush).

**Comment by tomachalek** [comment]
Removed *sk_SK* translation from *config.sample.xml* as the Slovak translation is not available.
